### PR TITLE
1330 Refine generate_spatial_bounding_box API

### LINKED ISSUE #1330: Combine compute_bounding_rect and generate_spatial_bounding_box
**Is your feature request related to a problem? Please describe.**
`generate_spatial_bounding_box` and `compute_bounding_rect` have similar functions, need to refine and combine them.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #1330 .

### Description
This PR combined `compute_bounding_rect` and `generate_spatial_bounding_box` APIs.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/transforms/croppad/array.py [hunk 1]
```diff
@@ -638,11 +637,23 @@ def __call__(self, img: np.ndarray, mode: Optional[Union[NumpyPadMode, str]] = N
...
-        bbox = [compute_bounding_rect(channel) for channel in img]
-        return np.stack(bbox, axis=0)
+        bbox_start = list()
+        bbox_end = list()
+
+        for channel in range(img.shape[0]):
+            start_, end_ = generate_spatial_bounding_box(img, select_fn=self.select_fn, channel_indices=channel)
+            bbox_start.append(start_)
+            bbox_end.append(end_)
+        return np.stack(bbox_start, axis=0), np.stack(bbox_end, axis=0)
```
Comment (wyli) [inline monai/transforms/croppad/array.py:659]: sorry I didn't make it clear, could this transform still return a 2D numpy array, in shape [`n_channels`, `(2*ndims)`]? currently it's 2 arrays of [`n_channels`, `ndims`]

File: monai/transforms/utils.py [hunk 2]
```diff
@@ -526,42 +535,21 @@ def generate_spatial_bounding_box(
     """
     data = img[[*(ensure_tuple(channel_indices))]] if channel_indices is not None else img
```
Comment (wyli) [inline monai/transforms/utils.py:536]: could be
```suggestion
    data = img[list(ensure_tuple(channel_indices))] if channel_indices is not None else img
```

File: monai/transforms/utils.py [hunk 3]
```diff
@@ -516,6 +516,15 @@ def generate_spatial_bounding_box(
     generate the spatial bounding box of foreground in the image with start-end positions.
     Users can define arbitrary function to select expected foreground from the whole image or specified channels.
     And it can also add margin to every dim of the bounding box.
+    The output format of the coordinates is:
+
+        [1st_spatial_dim_start, 1st_spatial_dim_end,
```
Comment (wyli) [inline monai/transforms/utils.py:521]: the output is now not in this format

File: monai/transforms/utils.py [hunk 4]
```diff
@@ -516,6 +516,15 @@ def generate_spatial_bounding_box(
...
+         2nd_spatial_dim_start, 2nd_spatial_dim_end,
+         ...,
+         Nth_spatial_dim_start, Nth_spatial_dim_end,]
+
+    The bounding boxes edges are aligned with the input image edges.
+    This function returns [-1, -1, ...] if there's no positive intensity.
```
Comment (wyli) [inline monai/transforms/utils.py:527]: this return value should be updated

File: monai/transforms/utils.py [hunk 5]
```diff
@@ -526,42 +535,21 @@ def generate_spatial_bounding_box(
...
+        dt = data.any(axis=ax)
         if not np.any(dt):
-            return np.asarray([-1] * len(bbox))
-        min_d = np.argmax(dt)
-        max_d = max(_binary_image.shape[di] - np.argmax(dt[::-1]), min_d + 1)
-        bbox[di * 2], bbox[di * 2 + 1] = min_d, max_d
-    return np.asarray(bbox)
+            return [-1] * ndim, [-1] * ndim
+
+        min_d = max(np.argmax(dt) - margin[di], 0)
```
Comment (wyli) [inline monai/transforms/utils.py:548]: do we support negative margins here? min max may need to swap

**Review by wyli — COMMENTED** [review]
thanks for the quick pr, I put some comments inline

File: monai/transforms/croppad/array.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/transforms/croppad/array.py:659]: OK, restore back to the original return format.
Thanks.

File: monai/transforms/utils.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/transforms/utils.py:536]: updated in the latest commit.
Thanks.

File: monai/transforms/utils.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/transforms/utils.py:527]: Updated in the latest commit.
Thanks.

File: monai/transforms/utils.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/transforms/utils.py:521]: Updated in the latest commit.
Thanks.

File: monai/transforms/utils.py (hunk 5 above)
Comment (Nic-Ma) [reply monai/transforms/utils.py:548]: We don't support negative margins.
I think `min_d` must be >= 0, so it's `max(..., 0)`, right? Which one you mean to swap?

Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks for your review.
I updated the PR according to your comments, could you please help review it again?

Thanks.

File: monai/transforms/utils.py (hunk 5 above)
Comment (wyli) [reply monai/transforms/utils.py:546]: thanks, currently `generate_spatial_bounding_box(np.asarray([[0, 0, 1, 1, 0]]), margin=0)` works as expected, but I'm not sure what would be the expected results for `generate_spatial_bounding_box(np.asarray([[0, 0, 1, 1, 0]]), margin=-2)`

Comment (Nic-Ma) [reply monai/transforms/utils.py:550]: Sure, I added value check for the margin.
Thanks.

**Review by wyli — APPROVED** [review]
thanks!

**Pull request merged**